compiler/rustc_codegen_llvm/src/coverageinfo/mapgen.rs

```diff
@@ -165,32 +165,32 @@ fn encode_mappings_for_function(
 
     let mut virtual_file_mapping = Vec::new();
     let mut mapping_regions = Vec::with_capacity(counter_regions.len());
-    let mut current_file_name = None;
-    let mut current_file_id = 0;
-
-    // Convert the list of (Counter, CodeRegion) pairs to an array of `CounterMappingRegion`, sorted
-    // by filename and position. Capture any new files to compute the `CounterMappingRegion`s
-    // `file_id` (indexing files referenced by the current function), and construct the
-    // function-specific `virtual_file_mapping` from `file_id` to its index in the module's
-    // `filenames` array.
+
+    // Sort the list of (counter, region) mapping pairs by region, so that they
+    // can be grouped by filename. Prepare file IDs for each filename, and
+    // prepare the mapping data so that we can pass it through FFI to LLVM.
     counter_regions.sort_by_key(|(_counter, region)| *region);
-    for (counter, region) in counter_regions {
-        let CodeRegion { file_name, start_line, start_col, end_line, end_col } = *region;
-        let same_file = current_file_name.is_some_and(|p| p == file_name);
-        if !same_file {
-            if current_file_name.is_some() {
-                current_file_id += 1;
-            }
-            current_file_name = Some(file_name);
-            debug!("  file_id: {} = '{:?}'", current_file_id, file_name);
-            let global_file_id = global_file_table.global_file_id_for_file_name(file_name);
-            virtual_file_mapping.push(global_file_id);
-        }
-        {
-            debug!("Adding counter {:?} to map for {:?}", counter, region);
+    for (local_file_id, counter_regions_for_file) in
+        (0u32..).zip(counter_regions.group_by(|(_, a), (_, b)| a.file_name == b.file_name))
```

Comment:
Nit: [`.enumerate()`](https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.enumerate) instead of `(0u32..).zip(...)`?

Reply:
I originally wrote this as enumerate, with a cast to `u32` later on.

But because the ID is supposed to be a `u32` (since that’s what LLVM wants), I wanted it to be a `u32` from the start to reflect that intent, which is why I switched to an explicit zip.

(Also, this way the `id as usize == len` assertion ends up doubling as an overflow check, though that was more of a happy coincidence.)

Reply:
I’m also thinking about removing the zip/enumerate entirely, and getting the local file ID directly from the vector length.

Right now it doesn’t make any difference, but in the future there will be reasons for the IDs to not be 1:1 with filenames, so that would save me from having to change this again later.